debian/civserver.6
.\" Freeciv - Copyright (C) 1996 - A Kjeldberg, L Gregersen, P Unold
.\"   This program is free software; you can redistribute it and/or modify
.\"   it under the terms of the GNU General Public License as published by
.\"   the Free Software Foundation; either version 2, or (at your option)
.\"   any later version.
.\"
.\"   This program is distributed in the hope that it will be useful,
.\"   but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"   GNU General Public License for more details.
.\"
.TH FREECIV 6 "July 1st 2004"
.SH NAME
civserver \- The server for the Freeciv game
.SH SYNOPSIS
.B civserver \
[ -d|--debug \fIlevel\fP ] \
[ -f|--file \fIsavegame\fP ] [ -g|--gamelog \fIgamelog\fP ] [ -h|--help ] \
[ -i|--info \fIinfo\fP ] [ -l|--log \fIlogfile\fP ] [ -m|--meta ] \
[ -M|--Metaserver \fIserver\fP ] [ -p|--port \fIport\fP ] \
[ -q|--quitidle \fIseconds\fP ] \
[ -r|--read \fIfile\fP ] [ -v|--version ]
.SH DESCRIPTION
Freeciv is a free turn-based multiplayer strategy game, in which each player
becomes the leader of a civilization, fighting to obtain the ultimate goal:
To become the greatest civilization.

Players of Civilization II\*R by Microprose\*R should feel at home, since one
aim of Freeciv is to have compatible rules. 

Freeciv is maintained by an international team of coders and enthusiasts, and is
easily one of the most fun and addictive network games out there!

This is the server program used to establish a Freeciv server - See also
.IR civclient (6)

This manual page only lists the command line arguments. For details
of the directives necessary to configure
.B civserver
see the Server Manual which can be found at
.UR
http://www.freeciv.org/
.UE
as well as the documentation which is part of the Freeciv distribution.
.SH OPTIONS
The following options are accepted on the command line of the server. They may
not be combined; that is, "civserver -fp savegame.sav 555" will not work,
instead you would need to enter "civserver -f savegame.sav -p 555". All options
have a short form (single hyphen and single letter) and a long form (double
hyphen and a complete word); their effects are identical.
.TP
.BI "-d, --debug"
Sets the amount of debugging information to be logged in the file named by the
.I -l
option. The number should be 0 for fatal errors only, 1 for fatal and normal
messages, 2 for fatal, normal, and debugging messages, or 3 for very verbose
output.
.TP
.BI "-f, --file"
Loads a saved game into the server before initialization, instead of starting a
new game. This is for reloading saved games, or for loading scenarios.
.TP
.BI "-g, --gamelog"
Defines a log file to be produced during processing. This is a game progress
log of game-related events recording the activities of the players.
.TP
.BI "-h, --help"
Prints out a description of the command line options and exits.
.TP
.BI "-i, --info"
Extra info to be sent to the metaserver.
.TP
.BI "-l, --log"
Defines a log file to be produced during processing. Use the
.I -d
option to set how much is logged.
.TP
.BI "-m, --meta"
Directs the server to communicate with a metaserver, a place where Freeciv
servers can publish their presence. Freeciv clients consult the metaserver to
offer users a selection of games to join.
.TP
.BI "-M, --Metaserver"
Sets the metaserver address.
.TP
.BI "-p, --port"
Specifies the TCP port number to which clients will connect; players must know
this number to be able to connect if they are not to use the default of 5555
decimal. You may need to use this if 5555 is not available for your use on your
system.
.TP
.BI "-q, --quitidle"
Quit if no players present for the given number of seconds.
.TP
.BI "-r, --read"
Specifies a file of server commands which the server will automatically read and
process, as if you had typed them in at the server's prompt.
.TP
.BI "-v, --version"
Causes the server to display its version number and exit.
.SH EXAMPLES
civserver --file oldgame.sav --port 2244

civserver -g gamelog -l logfile -r script -f oldgame.sav.gz
.SH COMMANDS
You may enter commands into the server at any time, either before or during the
running of a game. 

Type "help" or "help help" for starters.
.SH FILES
The Freeciv server requires the following files in the Freeciv data directory,
which on Debian systems is
.I /usr/share/games/freeciv/
by default:

\(bu
.I default/buildings.ruleset

\(bu
.I default/techs.ruleset

\(bu
.I default/units.ruleset

These are the default rule sets used for the game. Alternate sets of rules can
be installed with the option settings "techs", "units", and "buildings", each of
which provides a name to replace the directory name "default".
.SH ENVIRONMENT
The Freeciv server accepts these environment variables:
.TP
.BI FREECIV_CAPS
A string containing a list of "capabilities" provided by the server. The
compiled-in default should be correct for most purposes, but if you are familiar
with the capability facility in the source you may use it to enforce some
constraints between clients and server.
.TP
.BI FREECIV_PATH
A colon separated list of directories pointing to the
.B freeciv
data directory. By default freeciv looks in the following directories, in order,
for any data files: the current directory; the "data" subdirectory of the
current directory; the subdirectory ".freeciv" in the user's home directory; and
the directory where the files are placed by running "make install".
.TP
.BI FREECIV_LOCAL_ENCODING
The local encoding to be used.
.TP
.BI http_proxy
Set this variable accordingly when using a proxy.
.SH BUGS
Please report bugs to
.UR
http://rt.freeciv.org/
.UE
\.
.SH "MORE INFO"
The Freeciv homepage is located at
.UR
http://www.freeciv.org/
.UE
\.

Updates and new info is first posted there.
.SH AUTHORS
The Freeciv Team <freeciv-dev AT freeciv.org>.

This manpage was put together by Florian Ernst <florian_ernst AT gmx.net> using
the Server Manual and the comments in the sourcecode. Feel free to use it as
you wish.
.SH "SEE ALSO"
.IR civclient (6)
and the Server Manual on the Freeciv homepage.


ChangeLog

Sun, 05 Mar 2006 20:01:30 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/NEWS
	   M /branches/S2_0/common/version.h
	   M /branches/S2_0/doc/BUGS
	   M /branches/S2_0/version.in

	Bump version to 2.0.8.


Sun, 05 Mar 2006 19:47:21 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/common/packets.c

	Fix log compiler warning from PR#15713.


Sun, 05 Mar 2006 17:58:35 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/po/ca.po
	   M /branches/S2_0/po/cs.po
	   M /branches/S2_0/po/da.po
	   M /branches/S2_0/po/de.po
	   M /branches/S2_0/po/en_GB.po
	   M /branches/S2_0/po/es.po
	   M /branches/S2_0/po/et.po
	   M /branches/S2_0/po/fi.po
	   M /branches/S2_0/po/fr.po
	   M /branches/S2_0/po/he_IL.po
	   M /branches/S2_0/po/hu.po
	   M /branches/S2_0/po/it.po
	   M /branches/S2_0/po/ja.po
	   M /branches/S2_0/po/nl.po
	   M /branches/S2_0/po/no.po
	   M /branches/S2_0/po/pl.po
	   M /branches/S2_0/po/pt.po
	   M /branches/S2_0/po/pt_BR.po
	   M /branches/S2_0/po/ro.po
	   M /branches/S2_0/po/ru.po
	   M /branches/S2_0/po/sv.po
	   M /branches/S2_0/po/uk.po

	Run 'make update-po'.


Fri, 03 Mar 2006 18:42:00 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/common/packets.c

	Fix a potential crash in the packet reading code.

	Reported by Luigi Auriemma <aluigi@autistici.org> in PR#15762.


Wed, 01 Mar 2006 23:20:48 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/ai/ailog.h
	   M /branches/S2_0/client/climisc.h
	   M /branches/S2_0/client/gui-ftwl/gui_text.c
	   M /branches/S2_0/client/text.c
	   M /branches/S2_0/configure.ac
	   M /branches/S2_0/configure.in
	   M /branches/S2_0/server/console.h
	   M /branches/S2_0/server/plrhand.h
	   M /branches/S2_0/server/stdinhand.c
	   M /branches/S2_0/utility/astring.h
	   M /branches/S2_0/utility/fciconv.h
	   M /branches/S2_0/utility/ioz.h
	   M /branches/S2_0/utility/log.h
	   M /branches/S2_0/utility/registry.h
	   M /branches/S2_0/utility/shared.h
	   M /branches/S2_0/utility/support.h

	Use libutf8 to replace printf functions.  This happens automatically on
	windoes, or if --with-libutf8 is given on the configure line.  This replaces
	buggy printf functions on some systems that do not allow reordering of
	arguments (needed for localization support in some languages).

	Reported by Laurent Vogel <lvl@club-internet.fr>,
	Naoki Atagi <atagi@toki.waseda.jp>, and many others in PR#12932.  Patch by
	Christian Prochaska <cp.ml.freeciv.dev@googlemail.com>.


Wed, 01 Mar 2006 13:15:49 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/client/goto.c
	   M /branches/S2_0/client/options.c
	   M /branches/S2_0/client/options.h
	   M /branches/S2_0/common/aicore/path_finding.c
	   M /branches/S2_0/common/aicore/path_finding.h
	   M /branches/S2_0/common/aicore/pf_tools.c

	Backport: allow goto into the unknown.

	See PR#6977.


Mon, 27 Feb 2006 18:54:52 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/common/unit.c

	Fix a bug that allowed unlimited airlifts into a city.

	Reported by Jaroslav Libak <jarol1@seznam.cz> in PR#15582.


Sat, 21 Jan 2006 16:20:25 -0800  kauf:

	Changed paths:
	   M /branches/S2_0/server/auth.c
	   M /branches/S2_0/server/connecthand.c

	Fix bug where pconn->server.status wasn't updated.

	See PR#15098 for the gory details (and they are gory). Patch by me.


Sat, 21 Jan 2006 13:56:35 -0800  per:

	Changed paths:
	   M /branches/S2_0/po/es.po

	Additional Spanish translation by Jordi Negrevernis i Font
	<jnegrevernis@jnfprogramari.com>, posted in PR#14697.


Sat, 21 Jan 2006 13:54:07 -0800  per:

	Changed paths:
	   M /branches/S2_0/client/tilespec.c
	   M /branches/S2_0/server/score.c
	   M /branches/S2_0/server/unittools.c

	Silence silly gcc v4 warnings about possibly uninitialized
	variables (yeah, right).


Sun, 15 Jan 2006 16:04:49 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/client/gui-xaw

	svn:ignore


Fri, 13 Jan 2006 22:04:20 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/client/gui-xaw/cityrep.c
	   M /branches/S2_0/client/gui-xaw/dialogs.c
	   M /branches/S2_0/client/gui-xaw/repodlgs.c

	Backport: fix compiler warnings (and possible bugs) with wrongly terminated
	varargs function calls.


Sun, 08 Jan 2006 08:52:28 -0800  kauf:

	Changed paths:
	   M /branches/S2_0/server/auth.c
	   M /branches/S2_0/server/connecthand.c

	standardized logging to console for connection rejection. Add logging for
	all rejection cases.

	see PR#15101


Fri, 06 Jan 2006 21:35:48 -0800  kauf:

	Changed paths:
	   M /branches/S2_0/m4/auth.m4
	   M /branches/S2_0/manual/Makefile.am
	   M /branches/S2_0/server/Makefile.am

	attempt to find the correct LIBRARY_PATH for the mysql lib
	(for authentication) and put it in LDFLAGS. backport to S2_0

	see PR#14851.


Fri, 06 Jan 2006 21:21:59 -0800  kauf:

	Changed paths:
	   M /branches/S2_0/server/score.c

	remove the rest of the pre-post-processing needed to create
	the animated gifs for postgame. With this patch, you can simply do
	ppmtogif and then gifsicle. backport to S2_0

	see PR#14827.


Fri, 30 Dec 2005 10:38:29 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/server/score.c

	PR#14981 - don't always save ppm files.


Mon, 26 Dec 2005 15:17:38 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/client/civclient.c
	   M /branches/S2_0/client/clinet.c
	   M /branches/S2_0/client/gui-gtk-2.0/gui_main.c
	   M /branches/S2_0/client/packhand.c
	   M /branches/S2_0/common/dataio.c
	   M /branches/S2_0/common/dataio.h

	Fix compiler warnings in client.


Mon, 26 Dec 2005 14:52:28 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/common/aicore/cm.c
	   M /branches/S2_0/common/packets.c
	   M /branches/S2_0/server/sernet.c
	   M /branches/S2_0/utility/hash.c

	Backport: fix compiler warnings in server.


Mon, 26 Dec 2005 14:02:22 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/configure.in

	PR#12519: Missing socklen_t check in configure.in.


Mon, 26 Dec 2005 13:50:52 -0800  jdorje:

	Changed paths:
	   M /branches/S2_0/client/gui-gtk-2.0/citydlg.c

server/civserver.c
/**********************************************************************
 Freeciv - Copyright (C) 1996 - A Kjeldberg, L Gregersen, P Unold
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
***********************************************************************/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifndef WIN32_NATIVE
#include <signal.h>
#endif

#ifdef GENERATING_MAC  /* mac header(s) */
#include <Controls.h>
#include <Dialogs.h>
#endif

#ifdef WIN32_NATIVE
#include <windows.h>
#endif

#include "fciconv.h"
#include "fcintl.h"
#include "log.h"
#include "shared.h"
#include "support.h"
#include "version.h"

#include "console.h"
#include "meta.h"
#include "sernet.h"
#include "srv_main.h"

#include "civserver.h"

#ifdef GENERATING_MAC
static void Mac_options(int argc);  /* don't need argv */
#endif

/**************************************************************************
  Exit cleanly on SIGQUIT.
**************************************************************************/
#ifndef WIN32_NATIVE
static void signal_handler (int sig)
{
  if (sig == SIGQUIT) {
    fc_fprintf (stderr, _("\nServer quitting on SIGQUIT\n"));
    server_quit();
  }
}
#endif

/**************************************************************************
 Entry point for Freeciv server.  Basically, does two things:
  1. Parses command-line arguments (possibly dialog, on mac).
  2. Calls the main server-loop routine.
**************************************************************************/
int main(int argc, char *argv[])
{
  int inx;
  bool showhelp = FALSE;
  bool showvers = FALSE;
  char *option = NULL;

  /* Load win32 post-crash debugger */
#ifdef WIN32_NATIVE
# ifndef NDEBUG
  if (LoadLibrary("exchndl.dll") == NULL) {
#  ifdef DEBUG
    fprintf(stderr, "exchndl.dll could not be loaded, no crash debugger\n");
#  endif
  }
# endif
#endif

  /* initialize server */
  srv_init();

  /* parse command-line arguments... */

#ifdef GENERATING_MAC
  Mac_options(argc);
#endif
  /* no  we don't use GNU's getopt or even the "standard" getopt */
  /* yes we do have reasons ;)                                   */
  inx = 1;
  while (inx < argc) {
    if ((option = get_option("--file", argv, &inx, argc)))
      sz_strlcpy(srvarg.load_filename, option);
    else if (is_option("--help", argv[inx])) {
      showhelp = TRUE;
      break;
    } else if ((option = get_option("--log", argv, &inx, argc)))
      srvarg.log_filename = option;
    else if ((option = get_option("--gamelog", argv, &inx, argc)))
      srvarg.gamelog_filename = option;
    else if (is_option("--meta", argv[inx]))
      srvarg.metaserver_no_send = FALSE;
    else if ((option = get_option("--Metaserver", argv, &inx, argc))) {
      sz_strlcpy(srvarg.metaserver_addr, argv[inx]);
      srvarg.metaserver_no_send = FALSE;      /* --Metaserver implies --meta */
    } else if (is_option("--no-dns-lookup", argv[inx])) {
      srvarg.no_dns_lookup = TRUE;
    } else if ((option = get_option("--port", argv, &inx, argc))) {
      if (sscanf(option, "%d", &srvarg.port) != 1) {
	showhelp = TRUE;
	break;
      }
    } else if ((option = get_option("--bind", argv, &inx, argc))) {
      srvarg.bind_addr = option;
    } else if ((option = get_option("--read", argv, &inx, argc)))
      srvarg.script_filename = option;
    else if ((option = get_option("--quitidle", argv, &inx, argc))) {
      if (sscanf(option, "%d", &srvarg.quitidle) != 1) {
	showhelp = TRUE;
	break;
      }
    } else if (is_option("--exit-on-end", argv[inx])) {
      srvarg.exit_on_end = TRUE;
    } else if ((option = get_option("--debug", argv, &inx, argc))) {
      srvarg.loglevel = log_parse_level_str(option);
      if (srvarg.loglevel == -1) {
	srvarg.loglevel = LOG_NORMAL;
	showhelp = TRUE;
	break;
      }
#ifdef HAVE_AUTH
    } else if (is_option("--auth", argv[inx])) {
      srvarg.auth_enabled = TRUE;
    } else if (is_option("--Guests", argv[inx])) {
      srvarg.auth_allow_guests = TRUE;
    } else if (is_option("--Newusers", argv[inx])) {
      srvarg.auth_allow_newusers = TRUE;
#endif
    } else if (is_option("--Ppm", argv[inx])) {
      srvarg.save_ppm = TRUE;
    } else if ((option = get_option("--Serverid", argv, &inx, argc))) {
      sz_strlcpy(srvarg.serverid, option);
    } else if ((option = get_option("--saves", argv, &inx, argc))) {
      srvarg.saves_pathname = option;
    } else if ((option = get_option("--Require", argv, &inx, argc))) {
      sz_strlcpy(srvarg.required_cap, option);
    } else if (is_option("--version", argv[inx]))
      showvers = TRUE;
    else {
      fc_fprintf(stderr, _("Error: unknown option '%s'\n"), argv[inx]);
      showhelp = TRUE;
      break;
    }
    inx++;
  }

  if (showvers && !showhelp) {
    fc_fprintf(stderr, "%s %s\n", freeciv_name_version(), PEPSERVER_VERSION);
    exit(EXIT_SUCCESS);
  }
  con_write(C_VERSION, _("This is the server for %s"), freeciv_name_version());
  con_write(C_COMMENT, _("You can learn a lot about Freeciv at %s"),
	    WEBSITE_URL);

  if (showhelp) {
    fc_fprintf(stderr,
	       _("Usage: %s [option ...]\nValid options are:\n"), argv[0]);
#ifdef HAVE_AUTH
    fc_fprintf(stderr, _("  -a  --auth\t\tEnable server authentication.\n"));
    fc_fprintf(stderr, _("  -G  --Guests\t\tAllow guests to "
			 "login if auth is enabled.\n"));
    fc_fprintf(stderr, _("  -N  --Newusers\tAllow new users to "
			 "login if auth is enabled.\n"));
#endif
    fc_fprintf(stderr, _("  -b  --bind ADDR\tListen for clients on ADDR\n"));
#ifdef DEBUG
    fc_fprintf(stderr, _("  -d, --debug NUM\tSet debug log level (0 to 4,"
		      " or 4:file1,min,max:...)\n"));
#else
    fc_fprintf(stderr,
	       _("  -d, --debug NUM\tSet debug log level (0 to 3)\n"));
#endif
    fc_fprintf(stderr, _("  -f, --file FILE\tLoad saved game FILE\n"));
    fc_fprintf(stderr,
	       _("  -g, --gamelog FILE\tUse FILE as game logfile\n"));
    fc_fprintf(stderr,
	       _("  -h, --help\t\tPrint a summary of the options\n"));
    fc_fprintf(stderr, _("  -l, --log FILE\tUse FILE as logfile\n"));
    fc_fprintf(stderr, _("  -m, --meta\t\tNotify metaserver and "
			 "send server's info\n"));
    fc_fprintf(stderr, _("  -M, --Metaserver ADDR\tSet ADDR "
			 "as metaserver address\n"));
    fc_fprintf(stderr, _("  -n, --no-dns-lookup\tDo not lookup "
			 "the hostname of new connections\n"));

    fc_fprintf(stderr, _("  -p, --port PORT\tListen for clients on "
			 "port PORT\n"));
    fc_fprintf(stderr, _("  -q, --quitidle TIME\tQuit if no players "
			 "for TIME seconds\n"));
    fc_fprintf(stderr, _("  -e, --exit-on-end\t"
		      "When a game ends, exit instead of restarting\n"));
    fc_fprintf(stderr,
	       _("  -s, --saves DIR\tSave games to directory DIR\n"));
    fc_fprintf(stderr,
	       _("  -S, --Serverid ID\tSets the server id to ID\n"));
    fc_fprintf(stderr,
	     _("  -P, --Ppm\t\tSave ppms of the map when saving the game.\n"));
    fc_fprintf(stderr, _("  -r, --read FILE\tRead startup script FILE\n"));
    fc_fprintf(stderr, _("  -R, --Require CAP\tThe users will need this "
		      "capability(ies) to be able to play.\n"));
    fc_fprintf(stderr, _("  -v, --version\t\tPrint the version number\n"));
    fc_fprintf(stderr, _("Report bugs to <%s>.\n"), BUG_EMAIL_ADDRESS);
    exit(EXIT_SUCCESS);
  }

  /* disallow running as root -- too dangerous */
  dont_run_as_root(argv[0], "freeciv_server");

#ifndef WIN32_NATIVE
  if (signal (SIGQUIT, signal_handler) == SIG_ERR) {
    fc_fprintf (stderr, _("Failed to install SIGQUIT handler: %s\n"), mystrerror());
    exit (EXIT_FAILURE);
  }
#endif

  /* have arguments, call the main server loop... */
  srv_main();

  /* Technically, we won't ever get here. We exit via server_quit. */

  /* done */
  exit(EXIT_SUCCESS);
}

#ifdef GENERATING_MAC
/*************************************************************************
  generate an option dialog if no options have been passed in
*************************************************************************/
static void Mac_options(int argc)
{
#define HARDCODED_OPT
  /*temporary hack since GetNewDialog() doesn't want to work*/
#ifdef HARDCODED_OPT
  srvarg.log_filename="log.out";
  srvarg.loglevel=LOG_DEBUG;
#else
  if (argc == 0)
  {
    OSErr err;
    DialogPtr  optptr;
    Ptr storage;
    Handle ditl;
    Handle dlog;
    short the_type;
    Handle the_handle;
    Rect the_rect;
    short the_item, old_item=16;
    int done = false;
    Str255 the_string;
    /*load/init the stuff for the dialog*/
    storage =NewPtr(sizeof(DialogRecord));
    if (storage == 0)
    {
      exit(EXIT_FAILURE);
    }
    ditl=Get1Resource('DITL',200);
    if ((ditl == 0)||(ResError()))
    {
      exit(EXIT_FAILURE);
    }
    dlog=Get1Resource('DLOG',200);
    if ((dlog == 0)||(ResError()))
    {
      exit(EXIT_FAILURE);
    }
    /*make the dialog*/
    optptr=GetNewDialog(200, storage, (WindowPtr)-1L);
    /*setup the dialog*/
    err=SetDialogDefaultItem(optptr, 1);
    if (err != 0)
    {
      exit(EXIT_FAILURE);
    }
    /*insert default highlight draw code?*/
    err=SetDialogCancelItem(optptr, 2);
    if (err != 0)
    {
      exit(EXIT_FAILURE);
    }
    err=SetDialogTracksCursor(optptr, true);
    if (err != 0)
    {
      exit(EXIT_FAILURE);
    }
    GetDItem(optptr, 16/*normal radio button*/, &the_type, &the_handle, &the_rect);
    SetCtlValue((ControlHandle)the_handle, true);

    while(!done)/*loop*/
    {
      ModalDialog(0L, &the_item);/*don't feed 0 where a upp is expected?*/
      	/*old book sugests using OL(NIL) as the first argument, but
      	It doesn't include anything about UPPs either, so...*/
      switch (the_item)
      {
        case 1:
          done = true;
        break;
        case 2:
          exit(EXIT_SUCCESS);
        break;
        case 13:
          GetDItem(optptr, 13, &the_type, &the_handle, &the_rect);
          srvarg.metaserver_no_send=GetCtlValue((ControlHandle)the_handle);
          SetCtlValue((ControlHandle)the_handle, !srvarg.metaserver_no_send);
        break;
        case 15:
        case 16:
        case 17:
          GetDItem(optptr, old_item, &the_type, &the_handle, &the_rect);
          SetCtlValue((ControlHandle)the_handle, false);
          old_item=the_item;
          GetDItem(optptr, the_item, &the_type, &the_handle, &the_rect);
          SetCtlValue((ControlHandle)the_handle, true);
        break;
      }
    }
    /*now, load the dialog items into the corect variables interpritation*/
    GetDItem( optptr, 4, &the_type, &the_handle, &the_rect);
    GetIText( the_handle, (unsigned char *)srvarg.load_filename);
    GetDItem( optptr, 6, &the_type, &the_handle, &the_rect);
    GetIText( the_handle, (unsigned char *)srvarg.gamelog_filename);
    GetDItem( optptr, 8, &the_type, &the_handle, &the_rect);
    GetIText( the_handle, (unsigned char *)srvarg.log_filename);
    GetDItem( optptr, 12, &the_type, &the_handle, &the_rect);
    GetIText( the_handle, the_string);
    sscanf(the_string, "%d", srvarg.port);
    GetDItem( optptr, 10, &the_type, &the_handle, &the_rect);
    GetIText( the_handle, (unsigned char *)srvarg.script_filename);
    GetDItem(optptr, 15, &the_type, &the_handle, &the_rect);
    if(GetControlValue((ControlHandle)the_handle))
    {
      srvarg.loglevel=LOG_FATAL;
    }
    GetDItem(optptr, 16, &the_type, &the_handle, &the_rect);
    if(GetControlValue((ControlHandle)the_handle))
    {
      srvarg.loglevel=LOG_NORMAL;
    }
    GetDItem(optptr, 17, &the_type, &the_handle, &the_rect);
    if(GetControlValue((ControlHandle)the_handle))
    {
      srvarg.loglevel=LOG_VERBOSE;
    }
    DisposeDialog(optptr);/*get rid of the dialog after sorting out the options*/
    DisposePtr(storage);/*clean up the allocated memory*/
  }
#endif
#undef  HARDCODED_OPT
}
#endif
